src/sqlfluff/core/templaters/slicers/tracer.py

```diff
@@ -186,280 +214,319 @@ def slice_info_for_literal(self, length, prefix="") -> RawSliceInfo:
+        return JinjaTracer(
+            self.raw_str,
+            self.raw_sliced,
+            self.raw_slice_info,
+            self.sliced_file,
+            make_template,
+        )
+
+    def track_templated(
+        self, m_open: regex.Match, m_close: regex.Match, tag_contents: List[str]
+    ) -> RawSliceInfo:
+        """Compute tracking info for Jinja templated region, e.g. {{ foo }}."""
+        unique_alternate_id = self.next_slice_id()
+        open_ = m_open.group(1)
+        close_ = m_close.group(1)
+        # Here, we still need to evaluate the original tag contents, e.g. in
+        # case it has intentional side effects, but also return a slice ID
+        # for tracking.
+        alternate_code = (
+            f"\0{unique_alternate_id} {open_} " f"{''.join(tag_contents)} {close_}"
+        )
+        return self.make_raw_slice_info(unique_alternate_id, alternate_code)
+
+    def track_literal(self, raw: str) -> None:
+        """Set up tracking for a Jinja literal."""
+        self.raw_sliced.append(
+            RawFileSlice(
+                raw,
+                "literal",
+                self.idx_raw,
+            )
+        )
+        # Replace literal text with a unique ID.
+        self.raw_slice_info[self.raw_sliced[-1]] = self.slice_info_for_literal(
+            len(raw), ""
+        )
+        self.idx_raw += len(raw)
+
+    @staticmethod
+    def extract_block_type(tag_name, block_subtype):
+        """Determine block type."""
+        # :TRICKY: Syntactically, the Jinja {% include %} directive looks like
+        # a block, but its behavior is basically syntactic sugar for
+        # {{ open("somefile).read() }}. Thus, treat it as templated code.
+        if tag_name == "include":
+            block_type = "templated"
+        elif tag_name.startswith("end"):
+            block_type = "block_end"
+        elif tag_name.startswith("el"):
+            # else, elif
+            block_type = "block_mid"
+        else:
+            block_type = "block_start"
+            if tag_name == "for":
+                block_subtype = "loop"
+        return block_type, block_subtype
+
+    @staticmethod
+    def extract_tag_contents(
+        str_parts: List[str],
+        m_close: regex.Match,
+        m_open: regex.Match,
+        str_buff: str,
+    ) -> List[str]:
+        """Given Jinja tag info, return the stuff inside the braces.
+
+        I.e. Trim off the brackets and the whitespace.
+        """
+        if len(str_parts) >= 3:
+            # Handle a tag received as individual parts.
+            trimmed_parts = str_parts[1:-1]
+            if trimmed_parts[0].isspace():
+                del trimmed_parts[0]
+            if trimmed_parts[-1].isspace():
+                del trimmed_parts[-1]
+        else:
+            # Handle a tag received in one go.
+            trimmed_content = str_buff[len(m_open.group(0)) : -len(m_close.group(0))]
+            trimmed_parts = trimmed_content.split()
```

Comment:
Is there need for the If and the str_parts array at all, can the Else just be applied to both scenarios? (I guess its not expensive enough to matter about the splitting computation)

Reply:
Are you asking if we don't need to keep the individual `lex` results as separate strings?

The reason I did that is that Jinja doesn't just split on whitespace, so by keeping things as Jinja returns them, the code is more robust. Here's a test case for why:
```
{% set col= "col1" %}
SELECT {{ col }}
```

Note there's no space after `col`. Before switching to `str_parts`, we needed special handling since we might see any of:
```
{% set col= "col1" %}
{% set col = "col1" %}
{% set col ="col1" %}
```

This resulted in ugly, fragile code. Now it "just works"™️.

Reply:
I'll add the above as a code comment.

Reply:
str_buff is required, but we prefer to use str_parts because it makes better use of Jinja lexing.

(Not sure I'm fully understanding, though.)

If you feel like tinkering with it, the tests in `test/core/templaters/jinja_test.py` are pretty thorough.